platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/LongList.java

```diff
@@ -291,26 +390,109 @@ protected final void writeHeader(final FileChannel fc) throws IOException {
+     * will return {@link LongList#IMPERMISSIBLE_VALUE} for indices that
+     * are before {@code newMinValidIndex}.
+     * Also, a call to this method releases memory taken by unused byte buffers.
      *
      * @param newMinValidIndex minimal valid index of the list
      */
-    public void updateMinValidIndex(final long newMinValidIndex) {
-        // no-op
+    public final void updateMinValidIndex(final long newMinValidIndex) {
+        if (newMinValidIndex < 0) {
+            throw new IndexOutOfBoundsException("Min valid index " + newMinValidIndex + " must be non-negative");
+        }
+
+        if (newMinValidIndex >= size.get()) {
+            logger.warn("New min valid index cannot exceed current list size, returning.");
+            return;
+        }
+
+        // This is an optimization: we know that a chunk for this index must exist,
+        // so let's make sure that it's there
+        createOrGetChunk(newMinValidIndex);
+        shrinkIfNeeded(newMinValidIndex);
+        minValidIndex.set(newMinValidIndex);
+    }
+
+    /**
+     * Expand the available data storage if needed to allow storage of an item at newIndex
+     *
+     * @param newIndex the index of the new item we would like to add to storage
+     */
+    protected C createOrGetChunk(final long newIndex) {
+        size.getAndUpdate(oldSize -> newIndex >= oldSize ? (newIndex + 1) : oldSize);
+        final int chunkIndex = (int) (newIndex / numLongsPerChunk);
+        final C result = chunkList.get(chunkIndex);
+        if (result == null) {
+            final C newChunk = createChunk();
+            // set new chunk if it's not created yet, if it is - release the chunk immediately
+            // and use the one from the list
+            final C oldChunk = chunkList.compareAndExchange(chunkIndex, null, newChunk);
+            if (oldChunk == null) {
+                return newChunk;
+            } else {
+                fullChunkCleanup(newChunk);
+                return oldChunk;
+            }
+        } else {
+            return result;
+        }
+    }
+
+    /**
+     * Deletes values up to {@code newMinValidIndex}, releases memory chunks reserved for these
+     * values. Note that it takes {@code reservedBufferOffset} into account to decide if a chunk has
+     * to be released.
+     *
+     * @param newMinValidIndex new minimal valid index, left boundary of the list
+     */
+    private void shrinkIfNeeded(final long newMinValidIndex) {
+        final int firstValidChunkWithBuffer =
+                (int) Math.max((newMinValidIndex - reservedBufferLength) / numLongsPerChunk, 0);
+        final int firstChunkIndexToDelete = firstValidChunkWithBuffer - 1;
+        for (int i = firstChunkIndexToDelete; i >= 0; i--) {
+            final C chunk = chunkList.get(i);
+            if (chunk != null && chunkList.compareAndSet(i, chunk, null)) {
+                fullChunkCleanup(chunk);
+            }
```

Comment:
If `chunk` is null, break the loop? Is `compareAndSet()` really needed, or it can be an unconditional `set()`? Can the chunk array be modified in parallel on a different thread, e.g. from `put()`?

Reply:
>If chunk is null, break the loop? 

No, we can't break the loop. We have to traverse the whole chunk list because it may be sparse. That is, it may have null elements in in between non-null. 

> Is compareAndSet() really needed, or it can be an unconditional set()? 

Even though `updateMinValidIndex` is currently called from one thread at a time, I don't like the idea of conditional thread-safety. This method is designed (at least, that's the idea) to be fully thread-safe. And if we keep that in mind this `compareAndSet` call will prevent chunk duplicates in `LongListDisk`. The thing is that `fullChunkCleanup` allows to reuse the chunks (see `LongListDisk` implementation), and if it uses the same chunk twice when it reclaims the chunk, it will result in a data corruption.